lib/controllers.js

```diff
@@ -389,71 +415,58 @@ module.exports = function(rootDirectory, logger, indexDocument, errorDocument) {
-        (err, object, data) => {
-          if (err) {
-            if (indexDocument) {
-              keyName = path.join(keyName, indexDocument);
-              return fileStore.getObject(
-                req.bucket,
-                keyName,
-                (err, object, data) => {
-                  if (err) {
-                    return errorResponse(req, res, keyName);
-                  } else {
-                    return buildResponse(req, res, 200, object, data);
-                  }
-                }
-              );
-            } else {
-              return errorResponse(req, res, keyName);
-            }
-          }
-
-          const noneMatch = req.headers["if-none-match"];
-          if (
-            noneMatch &&
-            (noneMatch === JSON.stringify(object.md5) || noneMatch === "*")
-          ) {
-            return res.status(304).end();
-          }
-          const modifiedSince = req.headers["if-modified-since"];
-          if (modifiedSince) {
-            const time = new Date(modifiedSince);
-            const modifiedDate = new Date(object.modifiedDate);
-            if (time >= modifiedDate) {
-              return res.status(304).end();
-            }
-          }
+      const options = {};
+      if (/^bytes=/.test(req.headers["range"])) {
+        const [start, end] = req.headers["range"]
+          .replace("bytes=", "")
+          .split("-");
+        options.start = parseInt(start);
+        if (end) options.end = parseInt(end);
+      }
+      store.getObject(req.params.bucket, keyName, options, (err, object) => {
+        if (!object) {
+          if (!indexDocument) return errorResponse(req, res, keyName);
+          keyName = path.posix.join(keyName, indexDocument);
+          return store.getObject(req.params.bucket, keyName, (err, object) => {
+            if (!object) return errorResponse(req, res, keyName);
+            buildResponse(req, res, 200, object);
+          });
+        }
 
-          return buildResponse(
-            req,
-            res,
-            req.headers.range ? 206 : 200,
-            object,
-            data
-          );
+        const noneMatch = req.headers["if-none-match"];
+        if (
+          noneMatch &&
+          (noneMatch === object.metadata["etag"] || noneMatch === "*")
+        ) {
+          return res.status(304).end();
         }
-      );
+        const modifiedSince = req.headers["if-modified-since"];
+        if (new Date(modifiedSince) >= object.lastModifiedDate) {
+          return res.status(304).end();
+        }
+
+        return buildResponse(req, res, object.range ? 206 : 200, object);
+      });
     },
 
     putObject(req, res) {
-      if (req.query.uploadId) {
-        return putObjectMultipart(req, res);
-      }
-
       if (req.headers["x-amz-copy-source"]) {
-        handleCopyObject(req.params.key, req, res);
+        copyObject(req, res);
+      } else if (req.query.uploadId) {
+        putObjectMultipart(req, res);
```

Comment:
👍 Good job factoring out the `if (req.headers["x-amz-copy-source"])` branch from `putObjectMultipart`